crates/ra_hir/src/semantics.rs

```diff
@@ -70,6 +71,37 @@ impl<'db, DB: HirDatabase> Semantics<'db, DB> {
         Some(node)
     }
 
+    pub fn expand_hypothetical(
+        &self,
+        actual_macro_call: &ast::MacroCall,
+        hypothetical_call: &ast::MacroCall,
+        token_to_map: SyntaxToken,
+    ) -> Option<(SyntaxNode, SyntaxToken)> {
+        let macro_call =
+            self.find_file(actual_macro_call.syntax().clone()).with_value(actual_macro_call);
+        let sa = self.analyze2(macro_call.map(|it| it.syntax()), None);
+        let macro_call_id = macro_call
+            .as_call_id(self.db, |path| sa.resolver.resolve_path_as_macro(self.db, &path))?;
+        let macro_file = macro_call_id.as_file().macro_file().unwrap();
+        let (tt, tmap_1) =
+            hir_expand::syntax_node_to_token_tree(hypothetical_call.token_tree().unwrap().syntax())
+                .unwrap();
+        let range = token_to_map
+            .text_range()
+            .checked_sub(hypothetical_call.token_tree().unwrap().syntax().text_range().start())?;
+        let token_id = tmap_1.token_by_range(range)?;
+        let macro_def = hir_expand::db::expander(self.db, macro_call_id)?;
+        let (node, tmap_2) = hir_expand::db::parse_macro_with_arg(
+            self.db,
+            macro_file,
+            Some(std::sync::Arc::new((tt, tmap_1))),
+        )?;
+        let token_id = macro_def.0.map_id_down(token_id);
+        let range = tmap_2.range_by_token(token_id)?.by_kind(token_to_map.kind())?;
+        let token = algo::find_covering_element(&node.syntax_node(), range).into_token()?;
+        Some((node.syntax_node(), token))
+    }
+
```

Comment:
Neat! 
So what this function do is, use the original macro call with the `hypothetical_call` arguments. I see the limitation here: we don't want to *pollute* the salsa database such that we can't convert the `hypothetical_call` to an actual `MacroCallId`. So we have to expand it manually. 

One small nit: we might change the arguments to reflect it only replace the arguments :
```rust
    pub fn expand_hypothetical(
        &self,
        actual_macro_call: &ast::MacroCall,
        hypotheticall_args: &ast::TokenTree,
        token_to_map: SyntaxToken,
)
```

And I can imagine we could move some of the logic here to `hir_expand` but overall it is awesome !

Reply:
Good points, I'll move most of it to `hir_expand`, then I think it can be pretty contained there.